README.md
<!-- 
LOGO: https://user-images.githubusercontent.com/59066341/163708948-d5279779-6052-418d-ab08-a41f63491c3d.png
-->

<p align="center">
  <br>
  <img width="600" src="https://user-images.githubusercontent.com/59066341/163708948-d5279779-6052-418d-ab08-a41f63491c3d.png" alt="Widet Slider Logo">
  <br>
</p>

### Installing
- See the official installing guidline from [widget_slider/install](https://github.com/insolite-dev/widget_slider/wiki/Installing)

### Official Documentation (Wiki)
- Read the official documentation from [widget_slider/wiki](https://github.com/insolite-dev/widget_slider/wiki)

### Contributing
- For information regarding contributions, please refer to [CONTRIBUTING.md](https://github.com/insolite-dev/widget_slider/blob/main/CONTRIBUTING.md) file.


CHANGELOG.md
# v0.1.0 - (18/04/22)

## Very first release of `widget_slider`

 <img width="300" src="https://user-images.githubusercontent.com/59066341/163708948-d5279779-6052-418d-ab08-a41f63491c3d.png" alt="Widet Slider Logo">

- See the official installing guidline from [widget_slider/install](https://github.com/insolite-dev/widget_slider/wiki/Installing)
- Read the official documentation from [widget_slider/wiki](https://github.com/insolite-dev/widget_slider/wiki)
- For information regarding contributions, please refer to [CONTRIBUTING.md](https://github.com/insolite-dev/widget_slider/blob/main/CONTRIBUTING.md) file.



test/widget_slider_test.dart
//
// Copyright 2021-2022 present Insolite. All rights reserved.
// Use of this source code is governed by Apache 2.0 license
// that can be found in the LICENSE file.
//

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:widget_slider/widget_slider.dart';

void main() {
  late Widget horizontalSlider, verticalSlider;
  late SliderController controller;

  final items = List.generate(5, (index) => 'Test $index');

  setUpAll(() {
    controller = SliderController();

    horizontalSlider = WidgetSlider(
      controller: controller,
      itemCount: items.length,
      onMove: (_) {},
      itemBuilder: (context, index, activeIndex) {
        final value = index == activeIndex ? 'active' : 'non-active';
        return Text(value, key: Key(value));
      },
    );

    verticalSlider = WidgetSlider(
      itemCount: items.length,
      scrollDirection: Axis.vertical,
      onMove: (_) {},
      itemBuilder: (context, index, activeIndex) {
        return Text(index == activeIndex ? 'active' : 'non-active');
      },
    );
  });

  group('[WidgetSlider]', () {
    testWidgets('should build horizontal-widget properly', (tester) async {
      await tester.pumpWidget(
        MaterialApp(home: Scaffold(body: horizontalSlider)),
      );

      // Stable widgets.
      expect(find.byType(PageView), findsOneWidget);
      expect(find.byType(AnimatedBuilder), findsNWidgets(items.length + 3));
      expect(find.byType(Transform), findsNWidgets(8));
      expect(find.byType(Builder), findsNWidgets(items.length));
      expect(find.byType(SizedBox), findsNWidgets(2));

      // User provided.
      expect(find.byType(Text), findsNWidgets(2));

      // State actions.
      controller.moveToNext?.call();
      controller.moveToPrevious?.call();
      controller.moveTo?.call(items.length - 1);
    });

    testWidgets(
      'should build vertical-widget properly',
      (WidgetTester tester) async {
        await tester.pumpWidget(
          MaterialApp(home: Scaffold(body: verticalSlider)),
        );

        // Stable widgets.
        expect(find.byType(PageView), findsOneWidget);
        expect(find.byType(AnimatedBuilder), findsNWidgets(items.length + 3));
        expect(find.byType(Transform), findsNWidgets(8));
        expect(find.byType(Builder), findsNWidgets(items.length));
        expect(find.byType(SizedBox), findsNWidgets(2));

        // User provided.
        expect(find.byType(Text), findsNWidgets(2));
      },
    );
  });
}


lib/controller.dart
//
// Copyright 2021-2022 present Insolite. All rights reserved.
// Use of this source code is governed by Apache 2.0 license
// that can be found in the LICENSE file.
//

import 'package:flutter/material.dart';

/// The external controller for [WidgetSlider].
class SliderController {
  /// Duration value of sub-function's animations.
  ///
  /// If unset, defaults to the [milliseconds: 350].
  final Duration duration;

  /// Curve style of sub-function's animations.
  ///
  /// If unset, defaults to the [Curves.easeOut].
  final Curve curve;

  SliderController({
    this.duration = const Duration(milliseconds: 350),
    this.curve = Curves.easeOut,
  });

  /// Navigates to item at given index(i).
  ///
  /// If given index is exceeds min/max length, nothing will happen.
  Future<void> Function(int i)? moveTo;

  /// Navigates to next(current+1) item.
  ///
  /// If next(current+1) is exceeds max length, nothing will happen.
  Future<void> Function()? moveToNext;

  /// Navigates to previous(current-1) item.
  ///
  /// If back(current-1) is exceeds min length, nothing will happen.
  Future<void> Function()? moveToPrevious;
}


lib/widget_slider.dart
//
// Copyright 2021-2022 present Insolite. All rights reserved.
// Use of this source code is governed by Apache 2.0 license
// that can be found in the LICENSE file.
//

library widget_slider;

import 'package:flutter/material.dart';
import 'package:widget_slider/controller.dart';

export 'package:widget_slider/controller.dart';

/// A simple wrapper widget, used to generate size of [WidgetSlider].
class _FlexdSizedBox extends StatelessWidget {
  final Widget? child;
  final double? aspectRatio;
  final double? fixedSize;

  const _FlexdSizedBox({Key? key, this.child, this.aspectRatio, this.fixedSize})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (fixedSize != null) return SizedBox(height: fixedSize, child: child);
    return AspectRatio(aspectRatio: aspectRatio ?? 1, child: child);
  }
}

/// A simple carousel type component-slider widget for Flutter.
///
/// Example:
/// ```dart
/// WidgetSlider(
///   fixedSize: 300,
///   itemCount: images.length,
///   itemBuilder: (context, index, isActive) {
///     return Container(
///       decoration: const BoxDecoration(
///         image: DecorationImage(
///           image: NetworkImage('https://picsum.photos/200/300')
///         ),
///       ),
///     );
//    },
/// ),
/// ```
class WidgetSlider extends StatefulWidget {
  /// Length of items that'd be built via [itemBuilder].
  final int itemCount;

  /// Widget generator of the list. Generated widgets are represented as [PageView] pages.
  ///
  /// First index is each elements real index, but second index is active page's(item's) index.
  final Widget Function(BuildContext, int, int) itemBuilder;

  /// A function to execute on each page move.
  /// {i} represents the index of moved page.
  final Function(int i)? onMove;

  /// A external controller to manipulate widget from outside of it.
  ///
  /// next()    | navigates to next(current+1) item.
  /// back()    | navigates to back(current-1) item.
  /// moveTo(i) | navigates to item at given index(i).
  final SliderController? controller;

  /// The proximity between items.
  /// Value should be given as basic representation of percentages from (0 to 1) as
  /// | 0.1 -> 10%
  /// | 0.2 -> 20%
  /// ...
  ///
  /// If unset, defaults to the [0.5] -> 50%
  final double proximity;

  /// Decides differences of active & non-active item sizes .
  /// Value should be given as basic representation of percentages from (0 to 1) as
  /// | 0.1 -> 10%
  /// | 0.2 -> 20%
  /// ...
  ///
  ///           s:1
  ///  0.5   ╭────────╮   0.5
  /// ╭───╮  │        │  ╭───╮
  /// │ 6 │  │   42   │  │ 7 │
  /// ╰───╯  │        │  ╰───╯
  ///        ╰────────╯
  ///
  /// If unset, defaults to [0.5](non-active item will 50% smaller than active one).
  /// If set to [null], size distinction animation will be disabled.
  final double? sizeDistinction;

  /// The scroll direction of list widget.
  ///
  /// If unset, defaults to the [Axis.horizontal].
  final Axis scrollDirection;

  /// Ratio is of whole list widget.
  ///
  /// If unset, defaults to the -> [2].
  final double aspectRatio;

  /// Concrete fixed size for whole list widget.
  ///
  /// If not null, [aspectRatio]'s value will be ignored.
  final double? fixedSize;

  /// Animation curve style of widget's distinction.
  ///
  /// If unset defaults to the [Curves.easeOut].
  final Curve transformCurve;

  /// Reverse initializes list as reversed.
  /// [1, 2, 3] --> [3, 2, 1] (But remains at same page index(0)).
  ///
  /// If unset, defaults to the [false].
  final bool reverse;

  /// Enables/Disables snapping physics animation of widget.
  ///
  /// If unset, defaults to the [true](enabled).
  final bool pageSnapping;

  /// TODO: implement functionality to the widget.
  ///
  /// Enables/Disables infinite scrolling of list widget.
  /// If set, itemCount value will be ignored,
  /// Note: infinite scroll functionality could be enabled by setting [itemCount] to `null`.
  ///
  /// If unset, defaults to the [false](disabled).
  final bool infiniteScroll;

  /// Physics of the list widget.
  /// The physics are modified to snap to page boundaries using
  /// [PageScrollPhysics] prior to being used.
  ///
  /// If an explicit [ScrollBehavior] is provided to [scrollBehavior], the
  /// [ScrollPhysics] provided by that behavior will take precedence after
  /// [physics].
  ///
  /// Defaults to matching platform conventions.
  final ScrollPhysics? physics;

  /// Whether to add padding to both ends of the list.
  ///
  /// If this is set to true and [sizeDistinction] < 1.0, padding will be added
  /// such that the first and last child slivers will be in the center of
  /// the viewport when scrolled all the way to the start or end, respectively.
  ///
  /// If [sizeDistinction] >= 1.0, this property has no effect.
  ///
  /// If unset, defaults to the [true].
  final bool padEnds;

  const WidgetSlider({
    Key? key,
    required this.itemCount,
    required this.itemBuilder,
    this.onMove,
    this.controller,
    this.proximity = .5,
    this.sizeDistinction = 0.5,
    this.scrollDirection = Axis.horizontal,
    this.aspectRatio = 2,
    this.fixedSize,
    this.transformCurve = Curves.easeOut,
    this.reverse = false,
    this.pageSnapping = true,
    this.infiniteScroll = false,
    this.physics,
    this.padEnds = true,
  }) : super(key: key);

  @override
  // ignore: no_logic_in_create_state
  State<StatefulWidget> createState() => _WidgetSliderState(controller);
}

class _WidgetSliderState extends State<WidgetSlider> {
  _WidgetSliderState(SliderController? controller) {
    if (controller == null) return;
    _controller = controller;

    controller.moveTo = moveTo;
    controller.moveToNext = moveToNext;
    controller.moveToPrevious = moveToPrevious;
  }

  // Exact current page address, it's calculated from offsett.
  double page = 0;

  // Concrete current page adress, it's rounded variant of [page].
  // Used as [activeIndex] in [widget.itemBuilder].
  int activePage = 0;

  SliderController _controller = SliderController();
  late PageController pageController;

  @override
  void initState() {
    _reInitEssentials();
    super.initState();
  }

  @override
  void didUpdateWidget(covariant WidgetSlider oldWidget) {
    _reInitEssentials();
    super.didUpdateWidget(oldWidget);
  }

  void _reInitEssentials() {
    pageController = PageController(
      initialPage: activePage,
      viewportFraction: widget.proximity,
    );

    // Used to avoid [Page value is only available after content dimensions are established] error.
    // Taking page's value via listener causes no error. Instead of [pageController.page] we use [currentPage]
    pageController.addListener(() {
      if (pageController.page == null) return;
      page = pageController.page!;
    });
  }

  // Closure(callback) of controller's moveTo function.
  Future<void> moveTo(i) async {
    return await pageController.animateToPage(
      i,
      duration: _controller.duration,
      curve: _controller.curve,
    );
  }

  // Closure(callback) of controller's moveToNext function.
  Future<void> moveToNext() async {
    return await pageController.nextPage(
      duration: _controller.duration,
      curve: _controller.curve,
    );
  }

  // Closure(callback) of controller's moveToPrevious function.
  Future<void> moveToPrevious() async {
    return await pageController.previousPage(
      duration: _controller.duration,
      curve: _controller.curve,
    );
  }

  // Generates a valid scale for each item.
  double _generateScale(int index, Size size) {
    if (widget.sizeDistinction == null) return 1;

    final offset = (page - index).abs();
    final ratio = (1 - (widget.sizeDistinction! * offset)).clamp(0, 1);

    return widget.transformCurve.transform(ratio.toDouble());
  }

  @override
  Widget build(BuildContext context) {
    return _FlexdSizedBox(
      fixedSize: widget.fixedSize,
      aspectRatio: widget.aspectRatio,
      child: PageView.builder(
        controller: pageController,
        reverse: widget.reverse,
        itemCount: widget.itemCount,
        pageSnapping: widget.pageSnapping,
        physics: widget.physics,
        padEnds: widget.padEnds,
        scrollDirection: widget.scrollDirection,
        onPageChanged: (i) {
          // update static active page on each move.
          setState(() => activePage = i);

          widget.onMove?.call(i);
        },
        itemBuilder: (context, i) => AnimatedBuilder(
          animation: pageController,
          child: widget.itemBuilder(context, i, activePage),
          builder: (context, child) {
            final size = MediaQuery.of(context).size;
            final scale = _generateScale(i, size);

            return Transform.scale(
              scale: scale,
              child: Builder(
                builder: (context) {
                  if (widget.scrollDirection == Axis.horizontal) {
                    return SizedBox(child: child, width: scale * size.width);
                  }

                  return SizedBox(child: child, height: size.height);
                },
              ),
            );
          },
        ),
      ),
    );
  }
}


example/bubble_cards_example.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:widget_slider/widget_slider.dart';

void main() => runApp(const App());

class App extends StatelessWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: const Home(),
      theme: ThemeData.dark(),
    );
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final controller = SliderController();
  final indexes = List.generate(10, (index) => index+1);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: WidgetSlider(
          controller: controller,
          itemCount: indexes .length,
          onMove: (i) {/*...*/},
          itemBuilder: (context, index, activeIndex) {
            return CupertinoButton(
              padding: EdgeInsets.zero,
              onPressed: () async {
                if (index == activeIndex) return;
                await controller.moveTo?.call(index);
              },
              child: AnimatedContainer(
                duration: const Duration(milliseconds: 300),
                decoration: BoxDecoration(
                  color: index == activeIndex ? Colors.green : Colors.red,
                  shape: BoxShape.circle,
                  border: index == activeIndex
                      ? Border.all(color: Colors.white, width: 3)
                      : null,
                ),
                child: Center(
                  child: Text(
                    indexes[index].toString(),
                    style: const TextStyle(
                      fontWeight: FontWeight.bold,
                      color: Colors.white,
                    ),
                  ),
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}


example/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:widget_slider/widget_slider.dart';

void main() => runApp(const App());

class App extends StatelessWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: Home());
  }
}

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  final controller = SliderController(
    duration: const Duration(milliseconds: 600),
  );

  final images = const [
    'https://i.picsum.photos/id/816/200/300.jpg?hmac=4O5XSGjimzcjZYOXpVb_--v3rGzmS-3chmG2L1MS-mc',
    'https://i.picsum.photos/id/436/200/300.jpg?hmac=OuJRsPTZRaNZhIyVFbzDkMYMyORVpV86q5M8igEfM3Y',
    'https://i.picsum.photos/id/645/200/300.jpg?hmac=fiKW3Nj8r0CWJQY3S-kkeT8PAfvKhA8igd9GIRk41Yw',
    'https://i.picsum.photos/id/281/200/300.jpg?hmac=KCN8F5QTgxHdeQxLpZ5BOuUEVQEp8jAedlLSRERW7CY',
    'https://i.picsum.photos/id/816/200/300.jpg?hmac=4O5XSGjimzcjZYOXpVb_--v3rGzmS-3chmG2L1MS-mc',
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: WidgetSlider(
          fixedSize: 300,
          controller: controller,
          itemCount: images.length,
          itemBuilder: (context, index, activeIndex) {
            return CupertinoButton(
              onPressed: () async => await controller.moveTo?.call(index),
              child: Container(
                margin: const EdgeInsets.all(5),
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  image: DecorationImage(
                    fit: BoxFit.fill,
                    image: NetworkImage(images[index]),
                  ),
                  boxShadow: [
                    BoxShadow(
                      color: Colors.grey.shade50,
                      offset: const Offset(0, 8),
                      spreadRadius: 5,
                      blurRadius: 10,
                    ),
                    BoxShadow(
                      color: Colors.grey.shade100,
                      offset: const Offset(0, 8),
                      spreadRadius: 5,
                      blurRadius: 10,
                    ),
                  ],
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}
